lib/parser.ex

```diff
@@ -0,0 +1,205 @@
+defmodule Membrane.Ogg.Parser do
+  import Bitwise
+
+  @type packet :: %{payload: binary, track_id: integer, bos?: boolean, eos?: boolean}
+
+  @spec parse(binary, continued_packets :: %{}) ::
+          {parsed :: list(packet), unparsed :: binary, continued_packets :: %{}}
+  def parse(data, continued_packets) do
+    do_parse([], data, continued_packets)
+  end
+
+  @spec do_parse(acc :: list(packet), data :: binary, continued_packets :: %{}) ::
+          {parsed :: list(packet), data :: binary, continued_packets :: %{}}
+  defp do_parse(acc, data, continued_packets) do
+    case maybe_parse_page(data, continued_packets) do
+      {:error, :need_more_bytes} ->
+        {acc, data, continued_packets}
+
+      {:ok, unparsed, packets, continued_packets} ->
+        do_parse(acc ++ packets, unparsed, continued_packets)
+    end
+  end
+
+  @spec maybe_parse_page(page :: binary, continued_packets :: %{}) ::
+          {:ok, data :: binary, packets :: list(packet), continued_packets :: %{}}
+          | {:error, :need_more_bytes}
+  defp maybe_parse_page(initial_bytes, continued_packets) do
+    with {:ok, data, header_type, bitstream_serial_number, number_of_page_segments} <-
+           parse_header(initial_bytes),
+         {:ok, data, segment_table} <- parse_segment_table(data, number_of_page_segments),
+         :ok <- verify_length_and_crc(initial_bytes, segment_table) do
+      {data, packets, page_continued_packet} = parse_segments(data, segment_table)
```

Comment:
Let's use `rest` as it's more meaningful

```suggestion
      {packets, page_continued_packet, rest} = parse_segments(data, segment_table)
```